### ISSUE TITLE
Support for GPU checkpointing in nvproxy

### ISSUE THREAD

**Issue by cweld510** [google/gvisor#11095]
### Description

We're interested in some form of GPU checkpointing - is this something that the gvisor team plans on supporting at any point?

Generally, existing GPU checkpointing implementations described in papers like [Singularity](https://arxiv.org/pdf/2202.07848) or [Cricket](https://onlinelibrary.wiley.com/doi/epdf/10.1002/cpe.6474) intercept CUDA calls via `LD_PRELOAD`. Prior to a checkpoint, they record stateful calls in a log, which is stored at checkpoint time along with the contents of GPU memory. At restore time, GPU memory is reloaded and the log is replayed. Both frameworks have to do some of virtualization of device pointers as well.

It seems (perhaps naively) that a similar scheme might be possible within nvproxy, which already intercepts calls to the GPU driver. In theory, nvproxy could record a subset of calls made to the GPU driver and replay them at checkpoint-restore time, virtualizing file descriptors and device pointers as needed; and separately, support copying contents of GPU memory off the device to a file and back. 

This is clearly complex. I'm curious if you all believe it to be viable and plan on exploring the scheme described above, or a different one, at any point?

**Comment by EtiennePerot** [comment]
Have you looked at https://github.com/google/gvisor/issues/10478 (which I believe was filed by from one of your colleagues :))?
I believe `cuda-checkpoint` should work well within gVisor now that NVIDIA has fixed the issue described in that bug, and should allow GPU checkpointing to work in gVisor without the complexity of recording and replaying CUDA calls.

**Comment by cweld510** [comment]
Interesting, I assumed that NVIDIA hadn't fixed the issue since https://github.com/NVIDIA/cuda-checkpoint/issues/4 is still open, but honestly, I haven't tried running `cuda-checkpoint` again recently on pytorch within gvisor. I will do that.

**Comment by ayushr2** [comment]
I would recommend trying the latest driver (R565 I believe).

**Comment by cweld510** [comment]
Thanks! I'll reply back when I've had a chance to try the latest driver.

**Comment by tianyuzhou95** [comment]
> I would recommend trying the latest driver (R565 I believe).

As currently the latest driver gvisor support is 560.35.03, I have tried this with cuda-checkpoint and gvisor C/R. Unfortunately I still came across the same error during `runsc checkpoint` as @ayushr2 mentioned [here](https://github.com/google/gvisor/issues/10478#issuecomment-2127957702): `encoding error: can't save with live nvproxy clients`


> I believe cuda-checkpoint should work well within gVisor now that NVIDIA has fixed the issue described in that bug
It seems cuda-checkpoint itself has not updated, does the latest nvidia driver fix this bug?


Could you please give more information about how to make it(pytorch + cuda-checkpoint + gvisor C/R) work? Or is there any branch I could try on the latest driver(565)?


cc @amysaq2023 @btw616

**Comment by ayushr2** [comment]
@tianyuzhou95 It seems https://github.com/NVIDIA/cuda-checkpoint/issues/4 is still not fixed in any releases I have tried. This (pytorch apps not being able to be checkpointed) is a cuda-checkpoint bug, not a gVisor one. Could you follow up with NVIDIA about timeline?

**Comment by tianyuzhou95** [comment]
@ayushr2 Of course, I will follow Nvidia's fix for this. It looks like they plan to support it in early 2025. Thanks!

**Comment by ayushr2** [comment]
As per https://github.com/NVIDIA/cuda-checkpoint/issues/4#issuecomment-2616458251, NVML workloads (like Pytorch apps) should be checkpointable staring R570 drivers.

**Comment by ayushr2** [comment]
The recommended way of checkpointing CUDA applications in gVisor is by running the [`cuda-checkpoint`](https://github.com/NVIDIA/cuda-checkpoint) binary inside the gVisor sandbox on all CUDA processes. Then using `runsc checkpoint` to generate the checkpoint image.

On restore, use `runsc restore` to restore the sandbox and then run `cuda-checkpoint` on all the CUDA processes again to unfreeze them.
